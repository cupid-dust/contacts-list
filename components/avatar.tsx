'use client';
import Image from 'next/image';
import React, { useMemo } from 'react';
import { collapse } from '@growthops/ext-ts';

type Size = 'regular' | 'large' | 'small';

interface AvatarProps {
  className?: string;
  size: Size;
  image?: string;
  hasBorder?: boolean;
}

const sizeClasses: Record<Size, string> = {
  large: 'h-[88px] w-[88px] ',
  regular: 'w-10 h-10',
  small: 'w-6 h-6',
};

const useClasses = (
  size: Size,
  className: string | undefined,
  hasBorder?: boolean
) =>
  useMemo(
    () => ({
      root: collapse([
        sizeClasses[size],
        className ?? '',
        'overflow-hidden rounded-full',
        hasBorder ? 'border-2 border-white' : '',
      ]),
    }),
    [size, className, hasBorder]
  );

const Avatar = ({
  className,
  size = 'large',
  image = '/images/Default.jpg',
  hasBorder = false,
}: AvatarProps): JSX.Element => {
  const classes = useClasses(size, className, hasBorder);
  return (
    <div className={classes.root}>
      <Image src={image} alt="User Avatar" width={500} height={500} />
    </div>
  );
};

export type { Size };
export default Avatar;
