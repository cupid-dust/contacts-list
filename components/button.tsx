/* eslint-disable react/display-name */
'use client';

import React, { forwardRef, useMemo } from 'react';
import { isNil } from 'remeda';
import { collapse } from '@growthops/ext-ts';

type Variant = 'primary' | 'secondary' | 'tertiary';

type Size = 'small' | 'regular' | 'large';

type IconAlignment = 'left' | 'right' | 'center';

type Icon = {
  svg: React.ElementType;
  alignment?: IconAlignment;
  className?: string;
};

type BaseButtonProps = {
  label?: string;
  className?: string;
  variant?: Variant;
  size?: Size;
  icon?: Icon;
};

type SemanticButtonProps = BaseButtonProps &
  React.ComponentPropsWithoutRef<'button'>;

type LinkButtonProps = BaseButtonProps & React.ComponentPropsWithoutRef<'a'>;

const baseClasses =
  'inline-flex items-center space-x-2 transition-all ease-in-out duration-300 hover:cursor-pointer';

const variantClasses: Record<Variant, string> = {
  primary: 'bg-g-60 h-10 hover:bg-g-50 focus:bg-g-40 rounded-lg',
  secondary:
    'custom-gradient hover:hover-gradient focus:pressed-gradient rounded-full',
  tertiary: 'h-10 hover:bg-g-90 focus:bg-g-80 rounded-lg',
};

const sizeClasses: Record<Size, string> = {
  large: 'py-2 px-4 button-text ',
  regular: 'px-6 py-4 button-text',
  small: 'p-3 button-text',
};

const iconClasses: Record<Size, string> = {
  large: 'w-6',
  regular: 'w-5',
  small: 'w-4',
};

const generateIcon = (
  icon: Icon | undefined,
  alignment: IconAlignment,
  classes: string,
  alignmentAdjustmentClasses: string
) => {
  if (!isNil(icon) && (icon?.alignment ?? 'left') === alignment) {
    const IconComponent: any = icon.svg;
    return (
      <IconComponent
        className={collapse([classes, alignmentAdjustmentClasses])}
      />
    );
  }
};

const useClasses = (
  variant: Variant,
  size: Size,
  className: string | undefined,
  icon: Icon | undefined
) =>
  useMemo(
    () => ({
      root: collapse([
        baseClasses,
        variantClasses[variant],
        sizeClasses[size],
        className ?? '',
      ]),
      icon: collapse([iconClasses[size], icon?.className ?? '']),
      label: collapse([
        variant === 'secondary' ? 'md:inline-block hidden' : '',
      ]),
    }),
    [variant, size, className, icon?.className]
  );

const useIcons = (icon: Icon | undefined, classes: string) =>
  useMemo(
    () => ({
      left: generateIcon(icon, 'left', classes, '-ml-2'), // -ml-* here to visually weight the icon appropriately.
      right: generateIcon(icon, 'right', classes, '!-mr-2'), // !-mr-* here to visually weight the icon appropriately.
      center: generateIcon(icon, 'center', classes, 'm-0'), // to align icon in center.
    }),
    [icon, classes]
  );

const Semantic = ({
  label,
  variant = 'primary',
  size = 'regular',
  className,
  icon,
  ...intrinsicButtonProps
}: SemanticButtonProps): JSX.Element => {
  const classes = useClasses(variant, size, className, icon);
  const icons = useIcons(icon, classes.icon);

  return (
    <button className={classes.root} {...intrinsicButtonProps}>
      {icons.left}
      {icons.center}
      {!isNil(label) && <span className={classes.label}>{label}</span>}
      {icons.right}
    </button>
  );
};

const Link = forwardRef<HTMLAnchorElement, LinkButtonProps>(
  (
    {
      label,
      variant = 'primary',
      size = 'regular',
      className,
      icon,
      ...intrinsicAnchorProps
    }: LinkButtonProps,
    ref
  ): JSX.Element => {
    const classes = useClasses(variant, size, className, icon);
    const icons = useIcons(icon, classes.icon);

    return (
      <a
        ref={ref}
        className={classes.root}
        {...intrinsicAnchorProps}
        rel="noreferrer"
      >
        {icons.left}
        <span>{label}</span>
        {icons.right}
      </a>
    );
  }
);

const Button = {
  Semantic,
  Link,
};

export default Button;

export type {
  Variant,
  Size,
  IconAlignment,
  Icon,
  BaseButtonProps,
  SemanticButtonProps,
  LinkButtonProps,
};
