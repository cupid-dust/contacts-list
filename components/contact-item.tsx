import React from 'react';
import { UserInfo, Menu } from '@app/components';
import { HeadPhones, Mute } from '@app/components/svg';
import type { Contact } from '@app/types';

interface ContactItem {
  contact: Contact;
}

const ContactItem = ({ contact }: ContactItem): JSX.Element => {
  return (
    <div className="group flex items-center justify-between">
      <UserInfo
        name={contact?.name}
        phone={contact?.phone}
        image={contact?.image}
      />
      <div className="items-center gap-5 flex group-hover:visible invisible">
        <div className="cursor-pointer">
          <Mute />
        </div>
        <div className="cursor-pointer">
          <HeadPhones />
        </div>
        <Menu contact={contact} />
      </div>
    </div>
  );
};

export default ContactItem;
