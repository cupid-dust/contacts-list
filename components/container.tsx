'use client';

import { collapse } from '@growthops/ext-ts';
import React, { useMemo } from 'react';

type ContainerProps = {
  children: boolean | JSX.Element | Array<boolean | JSX.Element>;
  className?: string;
};

const Container = ({
  className = '',
  children,
}: ContainerProps): JSX.Element => {
  const classes = useMemo(
    () => ({
      root: collapse(['container', className]),
    }),
    [className]
  );

  return <div className={classes.root}>{children}</div>;
};

export default Container;

export type { ContainerProps };
