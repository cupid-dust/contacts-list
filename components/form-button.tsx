'use client';
import React from 'react';
import { Button } from '@app/components';
import { experimental_useFormStatus as useFormStatus } from 'react-dom';

const FormButton = (): JSX.Element => {
  const { pending } = useFormStatus();
  return (
    <Button.Semantic
      disabled={pending}
      label={pending ? 'loading...' : 'Done'}
      variant="primary"
      type="submit"
      size="large"
    />
  );
};
export default FormButton;
