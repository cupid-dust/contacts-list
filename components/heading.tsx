'use client';

import React from 'react';
import { collapse } from '@growthops/ext-ts';

type Variant = 'headline' | 'heading-one' | 'heading-two' | 'heading-three';

type SemanticElement = 'h1' | 'h2' | 'h3';

type HeadingProps = {
  variant: Variant;
  label: string;
  element?: SemanticElement;
  className?: string;
};

const Heading = ({
  variant,
  label,
  element = 'h1',
  className = '',
}: HeadingProps): JSX.Element => {
  const sharedProps = {
    className: collapse([variant, className]),
  };

  switch (element) {
    case 'h1':
      return <h1 {...sharedProps}>{label}</h1>;
    case 'h2':
      return <h2 {...sharedProps}>{label}</h2>;
    case 'h3':
      return <h3 {...sharedProps}>{label}</h3>;
    default:
      // Ensure exhaustive case handling.
      ((_: never): never => {
        throw new Error('Unhandled');
      })(element);
  }
};

export default Heading;

export type { Variant, SemanticElement, HeadingProps };
