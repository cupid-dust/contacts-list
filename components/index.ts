export { default as Button } from './button';
export { default as Container } from './container';
export { default as Heading } from './heading';
export { default as Input } from './input';
export { default as Text } from './text';
export { default as Avatar } from './avatar';
export { default as UserInfo } from './user-info';
export { default as ContactItem } from './contact-item';
export { default as Menu } from './menu';
export { default as Region } from './region';
export { default as Overlay } from './overlay';
export { default as FormButton } from './form-button';
