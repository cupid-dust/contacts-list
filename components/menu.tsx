'use client';
import { Menu as MenuWrapper, Transition } from '@headlessui/react';
import React, { Fragment } from 'react';
import { More, Settings, Heart, TrashBin } from '@app/components/svg';
import { experimental_useFormStatus as useFormStatus } from 'react-dom';
import { experimental_useFormState as useFormState } from 'react-dom';

import { useModal } from '@app/hooks';
import { removeContact } from '@app/lib';
import { Contact } from '@app/types';

const initialState = {
  message: null,
};

interface MenuProps {
  contact: Contact;
}

const Menu = ({ contact }: MenuProps): JSX.Element => {
  const [_, formAction] = useFormState(removeContact, initialState) as any;
  const { pending } = useFormStatus() as any;
  const { toggle } = useModal() as any;

  return (
    <div className="-mt-[6px]">
      <MenuWrapper as="div" className="relative inline-block text-left">
        <MenuWrapper.Button data-testid="more">
          <More />
        </MenuWrapper.Button>
        <Transition
          as={Fragment}
          enter="transition ease-out duration-100"
          enterFrom="transform opacity-0 scale-95"
          enterTo="transform opacity-100 scale-100"
          leave="transition ease-in duration-75"
          leaveFrom="transform opacity-100 scale-100"
          leaveTo="transform opacity-0 scale-95"
        >
          <MenuWrapper.Items className="overflow-hidden absolute right-0 mt-2 w-56 origin-top-right rounded-lg bg-g-80 ring-black ring-opacity-5 focus:outline-none z-10">
            <div>
              <MenuWrapper.Item>
                {({ active }) => (
                  <button
                    onClick={() => toggle(contact)}
                    className={`${
                      active ? 'bg-g-70 ' : 'bg-transparent'
                    } group   w-full    focus:bg-g-60 text-lead px-[10px] p-3`}
                  >
                    <span className="flex items-center gap-3">
                      <Settings className="fill-white/[56%]" /> Edit
                    </span>
                  </button>
                )}
              </MenuWrapper.Item>
              <MenuWrapper.Item>
                {({ active }) => (
                  <button
                    aria-disabled={pending}
                    className={`${
                      active ? 'bg-g-70 ' : 'bg-transparent'
                    } group gap-3 flex w-full items-center   focus:bg-g-60 text-lead px-[10px] p-3`}
                  >
                    <Heart className="fill-white/[56%]" /> Favourite
                  </button>
                )}
              </MenuWrapper.Item>
              <MenuWrapper.Item>
                {({ active }) => (
                  <form action={formAction}>
                    <input hidden name="id" value={contact?.id} />

                    <button
                      type="submit"
                      className={`${
                        active ? 'bg-g-70 ' : 'bg-transparent'
                      } group gap-3 flex w-full items-center   focus:bg-g-60 text-lead p-3`}
                    >
                      <TrashBin className="fill-white/[56%]" /> Remove
                    </button>
                  </form>
                )}
              </MenuWrapper.Item>
            </div>
          </MenuWrapper.Items>
        </Transition>
      </MenuWrapper>
    </div>
  );
};

export default Menu;
