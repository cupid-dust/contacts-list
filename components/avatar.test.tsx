import React from 'react';
import { expect, test } from '@jest/globals';
import renderer from 'react-test-renderer';
import { Avatar } from '@app/components';
import type { Size } from './avatar';

const sizes: Size[] = ['large', 'regular', 'small'];

test('snapshots â€” semantic', () => {
  for (const size of sizes) {
    const result = renderer
      .create(<Avatar size={size} image="/images/Default.jpg" />)
      .toJSON();

    expect(result).toMatchSnapshot();
  }
});
