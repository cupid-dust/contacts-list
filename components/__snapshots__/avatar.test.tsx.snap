// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots — semantic 1`] = `
<div
  className="h-[88px] w-[88px] overflow-hidden rounded-full"
>
  <img
    alt="User Avatar"
    data-nimg="1"
    decoding="async"
    height={500}
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="/_next/image?url=%2Fimages%2FDefault.jpg&w=1080&q=75"
    srcSet="/_next/image?url=%2Fimages%2FDefault.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2FDefault.jpg&w=1080&q=75 2x"
    style={
      {
        "color": "transparent",
      }
    }
    width={500}
  />
</div>
`;

exports[`snapshots — semantic 2`] = `
<div
  className="w-10 h-10 overflow-hidden rounded-full"
>
  <img
    alt="User Avatar"
    data-nimg="1"
    decoding="async"
    height={500}
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="/_next/image?url=%2Fimages%2FDefault.jpg&w=1080&q=75"
    srcSet="/_next/image?url=%2Fimages%2FDefault.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2FDefault.jpg&w=1080&q=75 2x"
    style={
      {
        "color": "transparent",
      }
    }
    width={500}
  />
</div>
`;

exports[`snapshots — semantic 3`] = `
<div
  className="w-6 h-6 overflow-hidden rounded-full"
>
  <img
    alt="User Avatar"
    data-nimg="1"
    decoding="async"
    height={500}
    loading="lazy"
    onError={[Function]}
    onLoad={[Function]}
    src="/_next/image?url=%2Fimages%2FDefault.jpg&w=1080&q=75"
    srcSet="/_next/image?url=%2Fimages%2FDefault.jpg&w=640&q=75 1x, /_next/image?url=%2Fimages%2FDefault.jpg&w=1080&q=75 2x"
    style={
      {
        "color": "transparent",
      }
    }
    width={500}
  />
</div>
`;
